### PR TITLE
feat: removed dependency on terraform batch jobs to apply integrations to keycloak

### PR DESCRIPTION
**Pull request by NithinKuruba**
Fixes below tickets
1. [SSOTEAM-1354](https://bcdevex.atlassian.net/browse/SSOTEAM-1354)
2. [SSOTEAM-1084](https://bcdevex.atlassian.net/browse/SSOTEAM-1084)
3. [SSOTEAM-1082](https://bcdevex.atlassian.net/browse/SSOTEAM-1082)
4. [SSOTEAM-1083](https://bcdevex.atlassian.net/browse/SSOTEAM-1083)

### CODE REVIEW

File: lambda/app/src/controllers/user.ts [hunk 1]
```diff
@@ -286,14 +286,10 @@ export const deleteStaleUsers = async (user: any) => {
...
-              const ghResult = await dispatchRequestWorkflow(rqst);
-
-              if (ghResult.status !== 204) {
-                throw Error('failed to create a workflow dispatch event');
-              }
 
               await disableIntegration(rqst.get({ plain: true, clone: true }));
-              await closeOpenPullRequests(rqst.id);
+              await dispatchRequestWorkflow(rqst);
+              //await closeOpenPullRequests(rqst.id);
```
Comment (jlangy) [inline lambda/app/src/controllers/user.ts:292]: We can probably remove this commented line

**Review by jlangy — COMMENTED** [review]
Looks good to me, just left a few small comments mostly about function names and old code. The only important comment was on the roles table, I think it might be good to add a unique constraint

File: app/pages/my-dashboard/integrations.tsx [hunk 2]
```diff
@@ -1,21 +1,84 @@
...
+  const router = useRouter();
   const [integration, setIntegration] = useState<Integration | null>(null);
   const [integrationCount, setIntegrationCount] = useState(1);
+  const [showModal, setShowModal] = useState(false);
+  const integrationFailedMessageModalId = 'integration-failed-modal';
+  const handleModalFailedMessageModal = async () => (window.location.hash = integrationFailedMessageModalId);
+
+  useEffect(() => {
+    // Check if the query parameter 'showModal' is present and set to 'true'
+    if (router.query.integrationFailedMessageModal === 'true') {
```
Comment (thegentlemanphysicist) [inline app/pages/my-dashboard/integrations.tsx:31]: will the query always have an integrationFaildMessageModal field? or should this be `query?.`

File: lambda/app/src/controllers/requests.ts [hunk 3]
```diff
@@ -465,17 +533,13 @@ export const deleteRequest = async (session: Session, user: User, id: number) =>
...
-
-    // disable the client while TF applying the changes
     await disableIntegration(current.get({ plain: true, clone: true }));
 
+    // Trigger workflow with empty environments to delete client
+    await dispatchRequestWorkflow(current);
+
     // Close any pr's if they exist
-    await closeOpenPullRequests(id);
+    //await closeOpenPullRequests(id);
```
Comment (thegentlemanphysicist) [inline lambda/app/src/controllers/requests.ts:542]: can this be deleted?

File: lambda/app/src/controllers/user.ts (hunk 1 above)
Comment (thegentlemanphysicist) [inline lambda/app/src/controllers/user.ts:292]: delete?

File: lambda/app/src/github.ts [hunk 4]
```diff
@@ -132,70 +133,64 @@ export const dispatchRequestWorkflow = async (integration: any) => {
...
-      await updatePlannedItems({
-        ids: [integration.id],
-        success: true,
-      });
-    }, 10000);
-  } catch (err) {
-    console.log(err);
-  }
-};
+// export const closeOpenPullRequests = async (id: number) => {
```
Comment (thegentlemanphysicist) [inline lambda/app/src/github.ts:148]: Is this staying here for future refactoring

File: lambda/shared/templates/restore-integration/index.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    subject: subjectHandler(data),
+    body: bodyHandler(data),
+  };
+};
+
+export const send = async (data: DataProps, rendered: RenderResult) => {
+  const { integration } = data;
+  const emails = await getIntegrationEmails(integration);
+  return sendEmail({
+    code: EMAILS.CREATE_INTEGRATION_APPLIED,
```
Comment (thegentlemanphysicist) [inline lambda/shared/templates/restore-integration/index.ts:35]: Is this supposed to use the same template as the create integration even though it's 'restore-integration' ?

Comment (NithinKuruba) [reply lambda/shared/templates/restore-integration/index.ts:35]: Renamed the dispatch function to `processIntegrationRequest` it does all the create/update/delete/restore ops

File: app/form-components/FormTemplate.tsx [hunk 6]
```diff
@@ -340,7 +340,10 @@ function FormTemplate({ currentUser, request, alert }: Props) {
 
         router.push({
           pathname: isAdmin ? '/admin-dashboard' : '/my-dashboard',
-          query: { id: formData.id },
+          query: {
+            id: data.id,
+            integrationFailedMessageModal: ['planFailed', 'applyFailed'].includes(data.status!),
```
[the remark this answers is not in the document]
Comment (NithinKuruba) [reply app/form-components/FormTemplate.tsx:345]: The condition specifically looks for failed statuses

File: lambda/app/src/github.ts [hunk 7]
```diff
@@ -132,70 +133,64 @@ export const dispatchRequestWorkflow = async (integration: any) => {
...
+//           repo: process.env.GH_REPO,
+//           issue_number: issue.number,
+//           state: 'closed',
+//         });
+//       }),
+//     );
+//   }
+// };
+
+// export const skipGithubActionStep = async (integrationId) => {
```
[the remark this answers is not in the document]
Comment (NithinKuruba) [reply lambda/app/src/github.ts]: Yes, removed now

File: lambda/app/src/github.ts [hunk 8]
```diff
@@ -88,7 +89,7 @@ export const buildGitHubRequestData = (baseData: IntegrationData) => {
   return baseData;
 };
 
-export const dispatchRequestWorkflow = async (integration: any) => {
+export const dispatchRequestWorkflow = async (integration: any, restore: boolean = false) => {
```
[the remark this answers is not in the document]
Comment (NithinKuruba) [reply lambda/app/src/github.ts]: Yeah we do not need github file anymore so moved all the code to requests controller file and renamed the function to `processIntegrationRequest`

File: lambda/db/src/migrations/2023.12.25T01.00.00.create-request-roles-table.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
+import { DataTypes } from 'sequelize';
+
+export const name = '2023.12.25T01.00.00.create-request-roles-table';
+
+// see https://sequelize.org/master/manual/naming-strategies.html
+export const up = async ({ context: sequelize }) => {
```
[the remark this answers is not in the document]
Comment (NithinKuruba) [reply lambda/db/src/migrations/2023.12.25T01.00.00.create-request-roles-table.ts:6]: Yes, that's a good suggestions I have included it.

File: lambda/app/src/controllers/user.ts (hunk 1 above)
Comment (NithinKuruba) [reply lambda/app/src/controllers/user.ts]: Removed

File: app/metadata/options.tsx [hunk 10]
```diff
@@ -1,9 +1,10 @@
 export const workflowStatusOptions = [
   { value: 'draft', label: 'Draft' },
   { value: 'submitted', label: 'Submitted' },
-  { value: 'pr', label: 'PR' },
-  { value: 'prFailed', label: 'PR Failed' },
+  // { value: 'pr', label: 'PR' },
```
[the remark this answers is not in the document]
Comment (NithinKuruba) [reply app/metadata/options.tsx]: yes removed it

File: app/pages/my-dashboard/integrations.tsx (hunk 2 above)
Comment (NithinKuruba) [reply app/pages/my-dashboard/integrations.tsx:30]: yes that's a good suggestion Jon, added it

File: lambda/app/src/controllers/requests.ts (hunk 3 above)
Comment (NithinKuruba) [reply lambda/app/src/controllers/requests.ts]: Removed

File: lambda/app/src/controllers/user.ts (hunk 1 above)
Comment (NithinKuruba) [reply lambda/app/src/controllers/user.ts]: removed

File: lambda/app/src/github.ts (hunk 4 above)
Comment (NithinKuruba) [reply lambda/app/src/github.ts]: Nope, removed it

**Review by jlangy — DISMISSED** [review]
looks great!

**Review by thegentlemanphysicist — DISMISSED** [review]

**Review by jlangy — DISMISSED** [review]

**Pull request merged**